File: benchmarks/fixtures/csharp-large/src/Api/Controllers/SubscriptionController.cs
using CSharpLargeApi.Application.DTOs;
using CSharpLargeApi.Application.Interfaces;
using CSharpLargeApi.Domain.Exceptions;
using CSharpLargeApi.Domain.ValueObjects;
using CSharpLargeApi.Infrastructure.Services;

namespace CSharpLargeApi.Api.Controllers;

/// <summary>
/// API controller for subscription management operations.
/// Provides endpoints for creating, renewing, and cancelling subscriptions.
/// </summary>
public class SubscriptionController
{
    private readonly SubscriptionService _subscriptionService;
    private readonly IPaymentService _paymentService;

    /// <summary>
    /// Initializes the controller with the subscription and payment services.
    /// </summary>
    public SubscriptionController(SubscriptionService subscriptionService, IPaymentService paymentService)
    {
        _subscriptionService = subscriptionService ?? throw new ArgumentNullException(nameof(subscriptionService));
        _paymentService = paymentService ?? throw new ArgumentNullException(nameof(paymentService));
    }

    /// <summary>
    /// Creates a new subscription for a user.
    /// POST /api/subscriptions
    /// </summary>
    public async Task<SubscriptionDto> Create(Guid userId, string planName, decimal monthlyPrice, string currency, string paymentMethodToken, CancellationToken cancellationToken = default)
    {
        var price = new Money(monthlyPrice, currency);
        var subscription = await _subscriptionService.CreateSubscription(userId, planName, price, cancellationToken);

        return new SubscriptionDto
        {
            Id = subscription.Id,
            UserId = subscription.UserId,
            PlanName = subscription.PlanName,
            MonthlyPrice = subscription.Price.Amount,
            Currency = subscription.Price.Currency,
            Status = subscription.Status.ToString(),
            NextRenewalDate = subscription.NextRenewalDate,
            CreatedAt = subscription.CreatedAt
        };
    }

    /// <summary>
    /// Renews a subscription and processes the renewal payment.
    /// Caller #3: SubscriptionController.RenewSubscription calls IPaymentService.ProcessPayment.
    /// POST /api/subscriptions/{id}/renew
    /// </summary>
    public async Task<SubscriptionDto> RenewSubscription(Guid subscriptionId, string paymentMethodToken, CancellationToken cancellationToken = default)
    {
        var subscription = await _subscriptionService.GetSubscriptionAsync(subscriptionId, cancellationToken);
        if (subscription is null)
        {
            throw new EntityNotFoundException("Subscription", subscriptionId);
        }

        // Process the renewal payment first
        // Caller #3: SubscriptionController.RenewSubscription calls ProcessPayment
        var payment = await _paymentService.ProcessPayment(
            subscription.UserId,
            subscription.Price,
            paymentMethodToken,
            cancellationToken);

        if (payment.Status != Domain.Enums.PaymentStatus.Completed)
        {
            throw new BusinessRuleException(
                "RenewalPaymentFailed",
                $"Renewal payment failed: {payment.FailureReason}");
        }

        // Then renew the subscription period
        var renewed = await _subscriptionService.RenewSubscription(subscriptionId, cancellationToken);

        return new SubscriptionDto
        {
            Id = renewed.Id,
            UserId = renewed.UserId,
            PlanName = renewed.PlanName,
            MonthlyPrice = renewed.Price.Amount,
            Currency = renewed.Price.Currency,
            Status = renewed.Status.ToString(),
            NextRenewalDate = renewed.NextRenewalDate,
            CreatedAt = renewed.CreatedAt
        };
    }

    /// <summary>
    /// Cancels a subscription.
    /// POST /api/subscriptions/{id}/cancel
    /// </summary>
    public async Task<SubscriptionDto> Cancel(Guid subscriptionId, CancellationToken cancellationToken = default)
    {
        var cancelled = await _subscriptionService.CancelSubscription(subscriptionId, cancellationToken);

        return new SubscriptionDto
        {
            Id = cancelled.Id,
            UserId = cancelled.UserId,
            PlanName = cancelled.PlanName,
            MonthlyPrice = cancelled.Price.Amount,
            Currency = cancelled.Price.Currency,
            Status = cancelled.Status.ToString(),
            NextRenewalDate = cancelled.NextRenewalDate,
            CreatedAt = cancelled.CreatedAt
        };
    }
}

File: benchmarks/fixtures/typescript-large/src/api/controllers/SubscriptionController.ts
import { Logger } from '../../shared/utils/Logger';
import { PaymentService } from '../../payments/services/PaymentService';
import { SubscriptionService } from '../../payments/services/SubscriptionService';
import { AuthGuard } from '../../auth/guards/AuthGuard';
import { ApiResponse } from '../../types/common';
import { PaymentResult } from '../../payments/types/PaymentTypes';
import { Subscription } from '../../payments/models/Subscription';
import { SubscriptionResponse, CreateSubscriptionRequest, CancelSubscriptionRequest } from '../../payments/dtos/SubscriptionDtos';
import { PaymentProcessor, BillingInterval } from '../../types/enums';
import { createMoney, Currency } from '../../types/money';
import { CryptoService } from '../../shared/utils/Crypto';

/** Controller for subscription endpoints */
export class SubscriptionController {
  private paymentService: PaymentService;
  private subscriptionService: SubscriptionService;
  private authGuard: AuthGuard;
  private logger: Logger;

  constructor(
    paymentService: PaymentService,
    subscriptionService: SubscriptionService,
    authGuard: AuthGuard,
  ) {
    this.paymentService = paymentService;
    this.subscriptionService = subscriptionService;
    this.authGuard = authGuard;
    this.logger = new Logger('SubscriptionController');
  }

  /** POST /subscriptions — create a new subscription */
  async create(authHeader: string, request: CreateSubscriptionRequest): Promise<ApiResponse<SubscriptionResponse>> {
    const user = this.authGuard.guard(authHeader);
    const amount = createMoney(request.amount, request.currency as Currency);

    const subscription = await this.subscriptionService.createSubscription(
      user.sub,
      request.planName,
      amount,
      request.billingInterval as BillingInterval,
    );

    return {
      success: true,
      data: {
        subscriptionId: subscription.id,
        planName: subscription.planName,
        status: subscription.status,
        billingInterval: subscription.billingInterval,
        amount: subscription.amount.amount,
        currency: subscription.amount.currency,
        currentPeriodStart: subscription.currentPeriodStart.toISOString(),
        currentPeriodEnd: subscription.currentPeriodEnd.toISOString(),
        nextBillingDate: subscription.nextBillingDate.toISOString(),
        createdAt: subscription.createdAt.toISOString(),
      },
      message: 'Subscription created',
      timestamp: new Date(),
    };
  }

  /**
   * POST /subscriptions/:id/renew
   * Manually triggers a subscription renewal.
   * CALLS processPayment (via SubscriptionService OR directly).
   */
  async renewSubscription(authHeader: string, subscriptionId: string): Promise<ApiResponse<PaymentResult>> {
    const user = this.authGuard.guard(authHeader);
    this.logger.info('Renewing subscription', { userId: user.sub, subscriptionId });

    const subscription = await this.subscriptionService.getSubscription(subscriptionId);
    const crypto = new CryptoService();

    const result = await this.paymentService.processPayment({
      userId: user.sub,
      amount: subscription.amount,
      processor: PaymentProcessor.STRIPE,
      description: `Manual renewal: ${subscription.planName}`,
      idempotencyKey: `manual_renew_${subscriptionId}_${crypto.generateToken(8)}`,
      metadata: { subscriptionId, planName: subscription.planName },
    });

    this.logger.info('Subscription renewal result', { subscriptionId, success: result.success });

    return {
      success: result.success,
      data: result,
      message: result.success ? 'Subscription renewed' : `Renewal failed: ${result.failureReason}`,
      timestamp: new Date(),
    };
  }

  /** POST /subscriptions/:id/cancel */
  async cancel(authHeader: string, subscriptionId: string, request: CancelSubscriptionRequest): Promise<ApiResponse<{ cancelled: boolean }>> {
    const user = this.authGuard.guard(authHeader);
    await this.subscriptionService.cancelSubscription(subscriptionId, request.reason);
    return {
      success: true,
      data: { cancelled: true },
      message: 'Subscription cancelled',
      timestamp: new Date(),
    };
  }
}
